### PR TITLE
Add constexpr and noexcept to SegmentId and ElementId where applicable

### PR DESCRIPTION
**Pull request by mar-celine**
Fixes issue #134

### CODE REVIEW

File: src/Domain/ElementId.hpp [hunk 1]
```diff
@@ -24,19 +24,21 @@ template <size_t VolumeDim>
 class ElementId {
  public:
   /// Default constructor needed for Charm++ serialization.
-  ElementId() = default;
+  constexpr ElementId() noexcept = default;
```
Comment (mar-celine) [inline src/Domain/ElementId.hpp]: This doesn't compile - figure out noexcept-ness of make_array as this is used to default construct a member variable of ElementId

File: src/Domain/SegmentId.hpp [hunk 2]
```diff
@@ -66,34 +68,33 @@ bool operator!=(const SegmentId& lhs, const SegmentId& rhs) noexcept;
...
+inline SegmentId SegmentId::id_of_child(Side side) const noexcept {
   // We cannot ASSERT on the maximum level because it's only known at runtime
   // and only known elsewhere in the code, not by SegmentId. I.e. SegmentId is
   // too low-level to know about this.
   // The child has twice as many segments as the parent, so for a particular
   // parent segment, there is both an upper and lower child segment.
   if (Side::Lower == side) {
     // left-shifting index_ by 1 gives us index*2.
-    return SegmentId(refinement_level_ + 1, index_ << 1);
+    return {refinement_level_ + 1, index_ << 1};
```
Comment (fmahebert) [inline src/Domain/SegmentId.hpp]: ditto

File: src/Domain/SegmentId.hpp [hunk 3]
```diff
@@ -66,34 +68,33 @@ bool operator!=(const SegmentId& lhs, const SegmentId& rhs) noexcept;
...
   }
   // left-shifting index_ by 1, followed by OR with 1, gives us index*2 +1.
-  return SegmentId(refinement_level_ + 1, 1 | (index_ << 1));
+  return {refinement_level_ + 1, 1 | (index_ << 1)};
```
Comment (fmahebert) [inline src/Domain/SegmentId.hpp]: ditto

File: src/Domain/SegmentId.hpp [hunk 4]
```diff
@@ -66,34 +68,33 @@ bool operator!=(const SegmentId& lhs, const SegmentId& rhs) noexcept;
...
-inline SegmentId SegmentId::id_of_parent() const {
-  ASSERT(0 != refinement_level_,
-         "Cannot call id_of_parent() on root refinement level!");
+inline SegmentId SegmentId::id_of_parent() const noexcept {
+  CASSERT(0 != refinement_level_,
+          "Cannot call id_of_parent() on root refinement level!");
   // right-shifting index_ by 1 gives us floor(index_/2).
   // The parent has half as many segments as the child.
-  return SegmentId(refinement_level_ - 1, index_ >> 1);
+  return {refinement_level_ - 1, index_ >> 1};
```
Comment (fmahebert) [inline src/Domain/SegmentId.hpp]: Was this bit magic used here to solve performance problems? If not, I'd favor the much more readable `floor(index_/2)` given in the comment.

File: src/Utilities/MakeArray.hpp [hunk 5]
```diff
@@ -28,8 +28,15 @@ namespace MakeArray_detail {
 /// is no risk of copy-after-move
 template <std::size_t size, typename T, std::size_t... indexes>
 SPECTRE_ALWAYS_INLINE constexpr std::array<std::decay_t<T>, size>
-make_array_impl(T&& value,
-                std::integer_sequence<size_t, indexes...> /* unused */) {
+make_array_impl(
+    T&& value,
+    std::integer_sequence<
+        size_t,
+        indexes...> /* unused */) noexcept(noexcept(std::array<typename std::
```
Comment (nilsdeppe) [inline src/Utilities/MakeArray.hpp]: `std::decay_t`

File: src/Utilities/MakeArray.hpp [hunk 6]
```diff
@@ -28,8 +28,15 @@ namespace MakeArray_detail {
 /// is no risk of copy-after-move
 template <std::size_t size, typename T, std::size_t... indexes>
 SPECTRE_ALWAYS_INLINE constexpr std::array<std::decay_t<T>, size>
-make_array_impl(T&& value,
-                std::integer_sequence<size_t, indexes...> /* unused */) {
+make_array_impl(
```
Comment (nilsdeppe) [inline src/Utilities/MakeArray.hpp]: Use `// clang-format off` and `// clang-format on`, then format this to something sane

File: src/Utilities/MakeArray.hpp [hunk 7]
```diff
@@ -28,8 +28,15 @@ namespace MakeArray_detail {
...
+    T&& value,
+    std::integer_sequence<
+        size_t,
+        indexes...> /* unused */) noexcept(noexcept(std::array<typename std::
+                                                                   decay<
+                                                                       T>::type,
+                                                               size>{
+    {(static_cast<void>(indexes), value)..., std::forward<T>(value)}})) {
   return std::array<typename std::decay<T>::type, size>{
```
Comment (nilsdeppe) [inline src/Utilities/MakeArray.hpp]: `typename std::decay<T>::type` to `std::decay_t<T>`

File: src/Utilities/MakeArray.hpp [hunk 8]
```diff
@@ -49,7 +57,12 @@ SPECTRE_ALWAYS_INLINE constexpr std::array<std::decay_t<T>, 0> make_array(
 /// helper function to make make_array<size_t> a simple call for users.
 template <std::size_t size, typename T>
 SPECTRE_ALWAYS_INLINE constexpr std::array<std::decay_t<T>, size> make_array(
-    std::integral_constant<std::size_t, size> /* unused */, T&& value) {
+    std::integral_constant<std::size_t, size> /* unused */,
+    T&& value) noexcept(noexcept(MakeArray_detail::
```
Comment (nilsdeppe) [inline src/Utilities/MakeArray.hpp]: override clang-format

File: src/Utilities/MakeArray.hpp [hunk 9]
```diff
@@ -75,7 +89,9 @@ SPECTRE_ALWAYS_INLINE constexpr std::array<std::decay_t<T>, size> make_array(
 template <typename T, typename... V, Requires<(sizeof...(V) > 0)> = nullptr>
 SPECTRE_ALWAYS_INLINE constexpr std::array<typename std::decay_t<T>,
                                            sizeof...(V) + 1>
-make_array(T&& t, V&&... values) {
+make_array(T&& t, V&&... values) noexcept(
+    noexcept(std::array<typename std::decay<T>::type, sizeof...(V) + 1>{
+        {std::forward<T>(t), std::forward<V>(values)...}})) {
   static_assert(
       cpp17::conjunction_v<std::is_same<std::decay_t<T>, std::decay_t<V>>...>,
```
Comment (nilsdeppe) [inline src/Utilities/MakeArray.hpp]: `tmpl2::flat_all_v<cpp17::is_same_v<std::decay_t<T>, std::decay_t<V>>`

File: tests/Unit/Domain/Test_SegmentId.cpp [hunk 10]
```diff
@@ -10,9 +10,6 @@
 #include <iostream>
 
 SPECTRE_TEST_CASE("Unit.Domain.SegmentId", "[Domain][Unit]") {
-  std::cout << noexcept(make_array<3>(0)) << std::endl;
```
Comment (nilsdeppe) [inline tests/Unit/Domain/Test_SegmentId.cpp]: ????

File: tests/Unit/Domain/Test_SegmentId.cpp [hunk 11]
```diff
@@ -10,9 +10,6 @@
 #include <iostream>
```
Comment (nilsdeppe) [inline tests/Unit/Domain/Test_SegmentId.cpp]: Write tests to check noexcept-ness of make_array inside the make array tests.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/Domain/Test_SegmentId.cpp [hunk 12]
```diff
@@ -7,7 +7,10 @@
 #include "Utilities/ConstantExpressions.hpp"
 #include "tests/Unit/TestHelpers.hpp"
 
+#include <iostream>
```
Comment (kidder) [inline tests/Unit/Domain/Test_SegmentId.cpp]: remove

Comment (kidder) [reply tests/Unit/Domain/Test_SegmentId.cpp]: this is flagged by the second build on Travis...

**Review by kidder — CHANGES_REQUESTED** [review]
t

File: tests/Unit/Domain/Test_SegmentId.cpp [hunk 13]
```diff
@@ -8,6 +8,7 @@
 #include "tests/Unit/TestHelpers.hpp"
 
 SPECTRE_TEST_CASE("Unit.Domain.SegmentId", "[Domain][Unit]") {
+
```
Comment (kidder) [inline tests/Unit/Domain/Test_SegmentId.cpp]: remove blank line

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**